docs/development/design/future-builder.rst

```diff
@@ -0,0 +1,230 @@
+* Re-write all ``readthedocs-sphinx-ext`` features to post-processsing HTML features
+* Reduce complexity maintained by Read the Docs' core team
+* Make Read the Docs responsible for Sphinx support and delegate other tools to the community
+* Eventually support upload pre-build docs
+* Allow us to add a feature with a defined contract without worry about breaking old builds
+* Introduce ``build.builder: 2`` config (does not install pre-defined packages) for these new features
+* Motivate users to migrate to ``v2`` to finally deprecate this magic by educating users
+
+
+Steps ran by the builder
+------------------------
+
+Read the Docs currently controls all the build process.
+Users are only allowed to modify very limited behavior by using a ``.readthedocs.yaml`` file.
+This drove us to implement features like ``sphinx.fail_on_warning``, ``submodules``, among others,
+at a high implementation and maintenance cost to the core team.
+Besides, this hasn't been enough for more advanced users that require more control over these commands.
+
+This document proposes to clearly define the steps the builder ran and allow users to override them
+depending on their needings:
+
+- Newbie user / simple platform usage: Read the Docs controls all the commands (current builder)
+- Intermediate user: ability to override one or more commands plus running pre/post hooks
+- Advanced user: controls *all the commands* executed by the builder
+
+The steps identified so far are:
+
+#. Checkout
+#. Expose project data via environment variables (\*)
+#. Create environment (virtualenv / conda)
+#. Install dependencies
+#. Build documentation
+#. Generate defined contract (``metadata.yaml``)
+#. Post-process HTML (\*)
+#. Upload to storage (\*)
+
+Steps marked with *(\*)* are managed by Read the Docs and can't be overwritten.
+
+
+Defined contract
+----------------
+
+Projects building on Read the Docs must provide a ``metadata.yaml`` file after running their last command.
+This file contains all the data required by Read the Docs to be able to add its integrations.
+If this file is not provided or malformed, Read the Docs will fail the build and stop the process
+communicating to the user that there was a problem with the ``metadata.yaml`` and we require them to fix the problem.
+
+.. note::
+
+   There is no restriction about how this file is generated
+   (e.g. generated with Python, Bash, statically uploaded to the repository, etc)
+   Read the Docs does not have control over it and it's only responsible for generating it when building with Sphinx.
+
+
+The following is an example of a ``metadata.yaml`` that is generated by Read the Docs when building Sphinx documentation:
+
+.. code:: yaml
+
+   # metadata.yaml
+   version: 1
+   tool:
+     name: sphinx
+     version: 3.5.1
+     builder: html
+   readthedocs:
+     html_output: ./_build/html/
+     pdf_output: ./_build/pdf/myproject.pdf
+     epub_output: ./_build/pdf/myproject.epub
```

Comment:
We could simplify this structure to be like the following to support other formats like `man` (see #4458):

```yaml
readthedocs:
  output:
    html:
    pdf:
    epub:
    man:
```

Besides, each of these keys could also be a list so we can support multiple PDF output (see #2045):

```yaml
readthedocs:
  output:
    pdf:
      - ./_build/pdf/myproject.pdf
      - ./_build/pdf/tutorial.pdf
      - ...
```

Reply:
+1 on both counts. I like the nested structure more and the list of files is a feature I've wanted for a while.